extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts

```diff
@@ -167,52 +168,261 @@ export class ChePluginManager {
+     */
+    private async getInstalledPlugins(filter: string): Promise<ChePlugin[]> {
+        // get list of all plugins
+        const rawPlugins: ChePluginMetadata[] = [];
+        for (let i = 0; i < this.registryList.length; i++) {
+            const registryPlugins: ChePluginMetadata[] = await this.chePluginService.getPlugins(this.registryList[i], filter);
+            registryPlugins.forEach(p => rawPlugins.push(p));
+        }
+
+        // group the plugins
+        const grouppedPlugins = this.groupPlugins(rawPlugins);
+
+        // prepare list of installed plugins without versions and repository URI
+        const installedPluginsInfo = this.getInstalledPluginsInfo();
+
+        const installedPlugins: ChePlugin[] = [];
+
+        // update `installed` field for all the plugin
+        // if the plugin is installed, we ned to set the proper version
+        grouppedPlugins.forEach(plugin => {
+            const publisherName = `${plugin.publisher}/${plugin.name}`;
+            installedPluginsInfo.forEach(info => {
+                if (info.publisherName === publisherName) {
+                    // set plugin is installed
+                    plugin.installed = true;
+                    // set intalled version
+                    plugin.version = info.version;
+
+                    installedPlugins.push(plugin);
+                }
+            });
+        });
+
+        return installedPlugins;
     }
 
-    async install(plugin: ChePluginMetadata): Promise<boolean> {
+    /**
+     * Returns list of installed plugins including installed version.
+     *
+     * Plugin should be without version and must not include plugin source.
+     *
+     * Plugin record
+     *     camel-tooling/vscode-apache-camel/0.0.14
+     * must be replaced on
+     *     camel-tooling/vscode-apache-camel
+     *
+     * Plugin record
+     *     https://raw.githubusercontent.com/vitaliy-guliy/che-theia-plugin-registry/master/plugins/eclipse-che/tree-view-sample-plugin/0.0.1/meta.yaml
+     * must be replaced on
+     *     eclipse-che/tree-view-sample-plugin
+     */
+    private getInstalledPluginsInfo(): { publisherName: string, version: string }[] {
+        // prepare the list of registries
+        // we need to remove the registry URI from the start of the plugin
+        const registries: string[] = [];
+        this.registryList.forEach(registry => {
+            let uri = registry.uri;
+            if (uri === this.defaultRegistry.uri) {
+                return;
+            }
+
+            if (uri.endsWith('.json')) {
+                uri = uri.substring(0, uri.lastIndexOf('/') + 1);
+            } else if (!uri.endsWith('/')) {
+                uri += '/';
+            }
+
+            registries.push(uri);
+        });
+
+        const plugins: { publisherName: string, version: string }[] = [];
+        this.installedPlugins.forEach(plugin => {
+            if (plugin.endsWith('/meta.yaml')) {
```

Comment:
What if a user gives full link to the default registry? Will it duplicate versions?

Reply:
Thanks for pointing me on this case. I did a note that I have to check the registry URI and don't allow the user to add already existing registry.